src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -524,13 +560,144 @@ class TypeComparer(initctx: Context) extends DotClass with ConstraintHandling {
       false
   }
 
+  /** Subtype test for the hk application `tp2 = tycon2[args2]`.
+   */
+  def compareHkApply2(tp1: Type, tp2: Type, tycon2: Type, args2: List[Type]): Boolean = {
```

Comment:
I'd suggest going with `tp2: HKApply` just for clarity (same in `compareHKApply1`)